doc/dirvish-expire.8
.\" Automatically generated by Pod::Man 2.1801 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DIRVISH-EXPIRE 1"
.TH DIRVISH-EXPIRE 1 "2009-02-04" "perl v5.10.0" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dirvish\-expire \- delete expired dirvish images
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdirvish-expire\fR\fI[\s-1OPTIONS\s0]\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Delete dirvish image trees or whole images that have expired.
.PP
Each image \fBsummary\fR file is checked for the \fBExpire:\fR field. If that
field indicates the image has expired \fBdirvish-expire\fR will delete
that image from the vault.
.PP
By default all subdirectories of all banks will be treated as vaults
and all directories therein except the one named \fBdirvish\fR will be
checked for summary files.
.PP
The removal of an image will have no effect on other images.
.PP
\&\fBDirvish-expire\fR will not delete an image unless it finds at least one
image in that branch that has an intact image tree and \fBStatus:
success\fR in the summary that is not expired.
.SH "OPTIONS"
.IX Header "OPTIONS"
Each option on the command line may be specified any number of times.
Those options that support lists in the config files will accumulate
all of their arguments otherwise each specification will override the
ones before.
.PP
Each option may be unambiguously abbreviated.
.PP
\&\fB\-\-time\fR \fItime_expression\fR
.PP
Execute as though \fItime_expression\fR were the current time.
.PP
\&\fITime_expression\fR is processed by \fB\f(BITime::Parsedate\fB\|(3pm)\fR so relative
time and date strings are permitted. See \fB\f(BITime::Parsedate\fB\|(3pm)\fR for
more details.
.PP
\&\fB\-\-tree\fR
.PP
Only delete the image tree, leave in place the rest of the image
directory with summary, log and any other image administrative files.
.PP
\&\fB\-\-vault\fR \fIvault\fR
.PP
Restrict expiration to the specified \fIvault\fR.
.PP
\&\fB\-\-no\-run\fR
.PP
Don't actually do anything. Just display what would have happened.
.PP
\&\fB\-\-quiet\fR
.PP
Run quietly, only report errors.
.PP
Normally \fBdirvish-expire\fR will report the images deleted.
.PP
\&\fB\-\-pidfile\fR
.PP
Specify an alternate location for the pidfile.
.SH "EXIT CODES"
.IX Header "EXIT CODES"
To facilitate further automation and integration of \fBdirvish-expire\fR
with other tools \fBdirvish-expire\fR provides rationalised exit codes.
The exit codes are range based. While the code for a specific error may
change from one version to another it will remain within the specified
range. So don't test for specific exit codes but instead test for a
range of values. To the degree possible higher value ranges indicate
more severe errors.
.IP "\fB0\fR" 10
.IX Item "0"
success
.IP "\fB200\-219\fR" 10
.IX Item "200-219"
An error was encountered in loading a configuration file.
.IP "\fB220\-254\fR" 10
.IX Item "220-254"
An error was detected in the configuration.
.IP "\fB255\fR" 10
.IX Item "255"
Incorrect usage.
.SH "FILES"
.IX Header "FILES"
\&\fB/etc/dirvish/master.conf\fR
.PP
alternate master configuration file.
.PP
\&\fB/etc/dirvish.conf\fR
.PP
master configuration file.
.PP
\&\fIbank/vault/image/\fR\fBsummary\fR
.PP
image creation summary.
.PP
\&\fIbank/vault/image/\fR\fBtree\fR
.PP
actual image of source directory tree.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.Vb 2
\& dirvish.conf(5)
\& Time::ParseDate(3pm)
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Dirvish was created by J.W. Schultz of Pegasystems Technologies.
.PP
Dirvish is now maintained by Keith Lofstrom at www.dirvish.org , with
the able assistance of many others.
.SH "BUGS"
.IX Header "BUGS"
\&\fBDirvish-expire\fR will walk the file hierarchy of all banks or the
specified vault looking for summary files. Anything non-dirvish in
there may cause excess file-walking.


debian/changelog
dirvish (1.3.2-r657-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Sun, 08 Feb 2009 00:41:43 +0100

dirvish (1.3.2-r652-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 04 Feb 2009 19:34:46 +0100

dirvish (1.3.2-r634-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 28 Jan 2009 17:46:52 +0100

dirvish (1.3.2-r632-1) unstable; urgency=low

  * Initial release

 -- Dominik Schulz <dominik.schulz@gauner.org>  Tue, 27 Jan 2009 16:18:14 +0100


dirvish-expire.pl
#!/usr/bin/perl
# dirvish-expire
# 1.3.X series
# Copyright 2005 by the dirvish project
# http://www.dirvish.org
#
# Last Revision   : $Rev: 660 $
# Revision date   : $Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $
# Last Changed by : $Author: tex $
# Stored as       : $HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-expire.pl $

#########################################################################
#                                                         				#
#	Licensed under the Open Software License version 2.0				#
#                                                         				#
#	This program is free software; you can redistribute it				#
#	and/or modify it under the terms of the Open Software				#
#	License, version 2.0 by Lauwrence E. Rosen.							#
#                                                         				#
#	This program is distributed in the hope that it will be				#
#	useful, but WITHOUT ANY WARRANTY; without even the implied			#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR				#
#	PURPOSE.  See the Open Software License for details.				#
#                                                         				#
#########################################################################
#
#----------------------------------------------------------------------------
# Revision information
#----------------------------------------------------------------------------

my %CodeID = (
    Rev    => '$Rev: 660 $'     ,
    Date   => '$Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $'    ,
    Author => '$Author: tex $'  ,
    URL    => '$HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-expire.pl $' ,
);

$VERSION =   $CodeID{URL};
$VERSION =~  s#^.*dirvish_##;  # strip off the front
$VERSION =~  s#\/.*##;         # strip off the rear after the last /
$VERSION =~  s#[_-]#.#g;       # _ or - to "."

#----------------------------------------------------------------------------
# Modules and includes
#----------------------------------------------------------------------------
use strict;
no strict qw(refs);
use warnings;

use Time::ParseDate;
use POSIX qw(strftime);
use Getopt::Long;
use Dirvish;

use File::Spec::Functions;

#----------------------------------------------------------------------------
# SIG Handler
#----------------------------------------------------------------------------
$SIG{TERM} = \&sigterm; # handles "kill <PID>"
$SIG{INT} = \&sigterm; # handles Ctrl+C or "kill -2 <PID>"

#----------------------------------------------------------------------------
# Initialisation
#----------------------------------------------------------------------------
my $Options = reset_options(\&usage, @ARGV);     # initialize the %$Options hash
load_master_config('f', $Options);

GetOptions($Options, qw(
    quiet!
    vault=s
    time=s
    pidfile=s
    tree!
    no-run|dry-run
    version
    help|?
    )) or &usage();

ref($$Options{vault}) and $$Options{vault} = undef;

my $expire_time = undef;
$$Options{time} and $expire_time = parsedate($$Options{time});
$expire_time ||= time;

my $pidfile = $$Options{pidfile};
$pidfile =~ s/\/dirvish[.]pid/\/dirvish-expire.pid/;

if(!check_pidfile($pidfile)) {
	print "$0 already running. Aborting at ".strftime('%H:%M:%S', localtime)." - PID: ".$$."\n";
	exit 255; # Abort
}

my @expires = ();
my %unexpired = ();

if ($$Options{vault})
{
    my $b;
    my $bank;
    for $b (@{$$Options{bank}})
    {
       	if (-d "$b/$$Options{vault}")
        {
            $bank = $b;
            last;
        }
    }
    $bank or seppuku 252, "Cannot find vault $$Options{vault}";
    Dirvish::find(catdir($bank, $$Options{vault}), $expire_time, $Options, \@expires, \%unexpired);
} else {
    for my $bank (@{$$Options{bank}})
    {
        Dirvish::find(catdir($bank), $expire_time, $Options, \@expires, \%unexpired);
    }
}

scalar(@expires) or exit 0;

if (!$$Options{quiet})
{
    printf "Expiring images as of %s\n",
        strftime('%Y-%m-%d %H:%M:%S', localtime($expire_time));
    $$Options{vault} and printf "Restricted to vault %s\n",
        $$Options{vault};
    print "\n";

    printf "%-15s %-15s %-16.16s  %s\n",
        qw(VAULT:BRANCH IMAGE CREATED EXPIRED);
}

# add nice and ionice to rm execution
my $basecmd = (($$Options{nice} && $$Options{nice} > -1 && &check_nice()) ? &check_nice()." -".$$Options{nice} : '');
$basecmd .= (($$Options{ionice} && $$Options{ionice} > -1 && &check_ionice()) ? &check_ionice()." -n".$$Options{ionice} : '');

# Iterate over the expired images and remove them
for my $expire (sort imsort @expires)
{
    my ($created, $expired);
    ($created = $$expire{created}) =~ s/:\d\d$//;
    ($expired = $$expire{expire}) =~ s/:\d\d$//;
    
    if (!$unexpired{$$expire{vault}}{$$expire{branch}})
    {
        printf "cannot expire %s:%s:%s No unexpired good images\n",
            $$expire{vault},
            $$expire{branch},
            $$expire{image};
        $$expire{status} =~ /^success/
            and ++$unexpired{$$expire{vault}}{$$expire{branch}};

        # By virtue of the sort order this will be the newest 
        # image so that older ones can be expired.
        next;
    }
    $$Options{quiet} or printf "%-15s %-15s %-16.16s  %s\n",
        $$expire{vault} . ':' .  $$expire{branch},
        $$expire{image},
        $created,
        $expired;

	print($basecmd." rm -rf $$expire{path}/tree\n") if($$Options{'no-run'} && !$$Options{quiet});
    system($basecmd." rm -rf $$expire{path}/tree") unless $$Options{'no-run'};
    
    $$Options{tree} and next;

	print($basecmd." rm -rf $$expire{path}\n") if($$Options{'no-run'} && !$$Options{quiet});
    system($basecmd." rm -rf $$expire{path}") unless $$Options{'no-run'};

}
exit 0;
#----------------------------------------------------------------------------
# Subs
#----------------------------------------------------------------------------
# Sort images
sub imsort {
	## WARNING:  don't mess with the sort order, it is needed so that if
	## WARNING:  all images are expired the newest will be retained.
	$$a{vault} cmp $$b{vault}
	  || $$a{branch} cmp $$b{branch}
	  || $$a{created} cmp $$b{created};
}
# Show usage information
sub usage
{
    my $message = shift(@_);

    length($message) and print STDERR $message, "\n\n";

    print STDERR <<EOUSAGE;
USAGE
	dirvish.expire OPTIONS
	
OPTIONS
	--time date_expression
	--[no]tree
	--vault vault_name
	--no-run
	--quiet
EOUSAGE

	exit 255;
}
# Remove the pidfile on exit
END {
	# remove the pidfile on exit
	remove_pidfile($pidfile);
}
# Handle SIGTERM (SIG-15)
sub sigterm
{
	print STDERR "Received SIGTERM. Aborting running backup ...";
	# kill childs - kill(TERM, -$$):
	use POSIX;
	my $cnt = kill(SIGTERM, -$$);
	no POSIX;
	print STDERR "Signaled $cnt processes in current processgroup";
	# quit
	exit;
}
#----------------------------------------------------------------------------
# EOF
#----------------------------------------------------------------------------
